
#Description:
This project aims to find talented individuals for the purpose of sourcing these candidates to technology companies. Finding talented candidates is not easy, for several reasons. The first reason is one needs to understand what the role is very well to fill in that spot, this requires understanding the client’s needs and what they are looking for in a potential candidate. The second reason is one needs to understand what makes a candidate shine for the role we are in search of. Third, where to find talented individuals is another challenge. The main goal of this project is to develop a machine learning-powered pipeline that could spot talented individuals, and rank them based on their fitness. This approach could save companies time and finally help them spot potential candidates that could fit the roles they are in search for. Generally, the sourcing part of candidates is based on some keywords such as “full-stack software engineer”, “engineering manager” or “aspiring human resources” based on the role companies are trying to fill in. After ranking fitting candidates, then employ a review procedure is required, to determine how good a fit they are. Indeed, our machine learning pipeline can re-rank the first list of ranked candidates based on keywords. Then, the list is expected to be re-ranked each time a candidate is starred.

#Steps:
The goal of this project is to predict how fit the candidate is based on their available information. I want to convey this through four stages:

Explore and analyse the data

Cleaning part

Modeling part

3.1. Rank candidates based on a given keyword

3.2. Re-rank the list whene some candidates are starred

In [37]:
pip install fasttext

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.4/73.4 kB 2.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached pybind11-2.13.6-py3-none-any.whl.metadata (9.5 kB)
Using cached pybind11-2.13.6-py3-none-any.whl (243 kB)
  Created wheel for fasttext: filename=fasttext-0.9.3-cp310-cp310-linux_x86_64.whl size=4296185 sha256=443f3f593a072823f68e03c0c6aa72c1b37bf385230d651c7324d52c8197e6a4
  Stored in directory: /root/.cache/pip/wheels/0d/a2/00/81db54d3e6a8199b829d58e02cec2ddb20ce3e59fad8d3c92a
Successfully built fasttext


In [38]:
pip install sentence_transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 245.3/245.3 kB 3.4 MB/s eta 0:00:00


In [39]:
#import required packages
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import nltk
import re
import string
from collections import Counter
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import RegexpTokenizer
from sklearn.metrics.pairwise import cosine_similarity
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from nltk import word_tokenize
import gensim
from gensim.scripts.glove2word2vec import glove2word2vec
from gensim.models import Word2Vec as w2v
from gensim.models import KeyedVectors
import fasttext.util
import csv
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE
from sentence_transformers import SentenceTransformer
import torch
import torch.nn as nn
import torch.optim as optim
import os
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

#1. Explore and analyse the dataset

In [40]:
import pandas as pd
data=pd.read_csv("potential-talents.csv").set_index('id')
data.head()

job_title  \
id                                                      
1   2019 C.T. Bauer College of Business Graduate (...   
2   Native English Teacher at EPIK (English Progra...   
3               Aspiring Human Resources Professional   
4              People Development Coordinator at Ryan   
5     Advisory Board Member at Celal Bayar University   

                               location connection  fit  
id                                                       
1                        Houston, Texas         85  NaN  
2                                Kanada      500+   NaN  
3   Raleigh-Durham, North Carolina Area         44  NaN  
4                         Denton, Texas      500+   NaN  
5                        İzmir, Türkiye      500+   NaN

In [41]:
#drop fit column
data.drop('fit',axis=1,inplace=True)
data.shape

(104, 3)

In [42]:
#count duplicated rows in the total dataset
data.loc[data.duplicated(), :]
print(data.duplicated().sum())

51


In [43]:
#drop duplicated rows in the total dataset
data.drop_duplicates(inplace=True)
data.shape

(53, 3)

In [44]:
#Verify if ther are some null values
data.isnull().sum()

job_title     0
location      0
connection    0
dtype: int64

In [45]:
#take a copy of the data
df=data.copy()

In [46]:
df.job_title.value_counts()

job_title
Aspiring Human Resources Professional                                                                                    2
2019 C.T. Bauer College of Business Graduate (Magna Cum Laude) and aspiring Human Resources professional                 1
Lead Official at Western Illinois University                                                                             1
Senior Human Resources Business Partner at Heil Environmental                                                            1
Aspiring Human Resources Professional | An energetic and Team-Focused Leader                                             1
HR Manager at Endemol Shine North America                                                                                1
Human Resources professional for the world leader in GIS software                                                        1
RRP Brand Portfolio Executive at JTI (Japan Tobacco International)                                                       1
Information Systems Specialist and Programmer with a love for data and organization.                                     1
Bachelor of Science in Biology from Victoria University of Wellington                                                    1
Human Resources Management Major                                                                                         1
Director Human Resources  at EY                                                                                          1
Undergraduate Research Assistant at Styczynski Lab                                                                       1
Seeking employment opportunities within Customer Service or Patient Care                                                 1
Liberal Arts Major. Aspiring Human Resources Analyst.                                                                    1
Admissions Representative at Community medical center long beach                                                         1
Seeking Human  Resources Opportunities. Open to travel and relocation.                                                   1
Student at Westfield State University                                                                                    1
Student at Indiana University Kokomo - Business Management - \nRetail Manager at Delphi Hardware and Paint               1
Student                                                                                                                  1
Seeking Human Resources Position                                                                                         1
Aspiring Human Resources Manager | Graduating May 2020 | Seeking an Entry-Level Human Resources Position in St. Louis    1
Human Resources Generalist at Loparex                                                                                    1
Business Intelligence and Analytics at Travelers                                                                         1
Always set them up for Success                                                                                           1
Junior MES Engineer| Information Systems                                                                                 1
Human Resources Generalist at Schwan's                                                                                   1
Native English Teacher at EPIK (English Program in Korea)                                                                1
Seeking Human Resources Opportunities                                                                                    1
People Development Coordinator at Ryan                                                                                   1
Advisory Board Member at Celal Bayar University                                                                          1
Aspiring Human Resources Specialist                                                                                      1
Student at Humber College and Aspiring Human Resources

As we can dedicate, these texts need some preprocessing, including removing stopwords, uppercases, abbreviations, applying lemmatization, removing digits, and so on. So, it's time to apply the preprocessing to the text.

#2. Cleaning the dataset

Define preprocessing function to clean the text

In [47]:
#define our preprocessing function
def preprocess_text(text):
    all_reviews=list()
    # Tokenise words while ignoring punctuation
    tokeniser = RegexpTokenizer(r'\w+')
    tokens = tokeniser.tokenize(text)
    # Lowercase and lemmatise , pos='v' means verb
    lemmatiser = WordNetLemmatizer()
    lemmas = [lemmatiser.lemmatize(token.lower(),pos='v') for token in tokens]
    # Remove stopwords
    keywords= [lemma for lemma in lemmas if lemma not in stopwords.words('english')]
    x=' '.join(keywords)
    x=' '.join([w for w in x.split() if len(w)>1])
    x= re.sub('\w*\d\w*','', x)#remove digits and/or words contain digit
    x= re.sub("türkiye","turkey",x)#this cleans the location column
    x= re.sub("kanada","canada",x)#this cleans the location column
    x= re.sub("amerika birleşik devletleri","united states america",x)#this cleans the location column
    all_reviews.append(x)
    return x

In [48]:
#explore some english stopwords
print('len of stop word in english:',len(stopwords.words('english')))
stopwords.words('english')[:5]

len of stop word in english: 179


['i', 'me', 'my', 'myself', 'we']

In [49]:
#apply preprocessing on the job_title feature
df['job_title'] = df.job_title.apply(preprocess_text)

In [50]:
#replace abreviations by there real texts
df.replace({'job_title' : { 'chro' : 'chief human resources officer', 'svp' : 'senior vice president'
        ,'gphr' : 'global professional in human resources','hris' : 'human resources management system'
        , 'csr' : 'corporate social responsibility', 'sphr' : 'strategic and policy-making certification'
        , 'hr' : 'human resources'}}, regex=True, inplace=True)

In [51]:
#Example after text preprocessing
print(data.iloc[0]['job_title'])
df.iloc[0]['job_title']

2019 C.T. Bauer College of Business Graduate (Magna Cum Laude) and aspiring Human Resources professional


' bauer college business graduate magna cum laude aspire human resources professional'

In [52]:
df.head()

job_title  \
id                                                      
1    bauer college business graduate magna cum lau...   
2   native english teacher epik english program korea   
3                 aspire human resources professional   
4                 people development coordinator ryan   
5        advisory board member celal bayar university   

                               location connection  
id                                                  
1                        Houston, Texas         85  
2                                Kanada      500+   
3   Raleigh-Durham, North Carolina Area         44  
4                         Denton, Texas      500+   
5                        İzmir, Türkiye      500+

In [53]:
df.job_title.value_counts()

job_title
aspire human resources professional                                                                                                                                                                                                  2
 bauer college business graduate magna cum laude aspire human resources professional                                                                                                                                                 1
lead official western illinois university                                                                                                                                                                                            1
senior human resources business partner heil environmental                                                                                                                                                                           1
aspire human resources professional energetic team focus leader                                                                                                                                                                      1
human resources manager endemol shine north america                                                                                                                                                                                  1
human resources professional world leader gi software                                                                                                                                                                                1
rrp brand portfolio executive jti japan tobacco international                                                                                                                                                                        1
information systems specialist programmer love data organization                                                                                                                                                                     1
bachelor science biology victoria university wellington                                                                                                                                                                              1
human resources management major                                                                                                                                                                                                     1
director human resources ey                                                                                                                                                                                                          1
undergraduate research assistant styczynski lab                                                                                                                                                                                      1
seek employment opportunities within customer service patient care                                                                                                                                                                   1
liberal arts major aspire human resources analyst                                                                                                                                                                                    1
admissions representative community medical center long beach                                                                                                                                                                        1
seek human resources opportunities open travel relocation                                                                                                                                                                            1
student westfield state university                             

In [54]:
#apply preprocessing on the job_title feature
df['job_title'] = df.job_title.apply(preprocess_text)

In [55]:
df.job_title.value_counts()

job_title
aspire human resources professional                                                                                                                                                                                         2
bauer college business graduate magna cum laude aspire human resources professional                                                                                                                                         1
lead official western illinois university                                                                                                                                                                                   1
senior human resources business partner heil environmental                                                                                                                                                                  1
aspire human resources professional energetic team focus leader                                                                                                                                                             1
human resources manager endemol shine north america                                                                                                                                                                         1
human resources professional world leader gi software                                                                                                                                                                       1
rrp brand portfolio executive jti japan tobacco international                                                                                                                                                               1
information systems specialist programmer love data organization                                                                                                                                                            1
bachelor science biology victoria university wellington                                                                                                                                                                     1
human resources management major                                                                                                                                                                                            1
director human resources ey                                                                                                                                                                                                 1
undergraduate research assistant styczynski lab                                                                                                                                                                             1
seek employment opportunities within customer service patient care                                                                                                                                                          1
liberal arts major aspire human resources analyst                                                                                                                                                                           1
admissions representative community medical center long beach                                                                                                                                                               1
seek human resources opportunities open travel relocation                                                                                                                                                                   1
student westfield state university                                                                                                                                                                                      

Now, as a last step, let's see the frequency of word's occurrence in our job titles

In [56]:
words_counts = Counter()
for i in df.job_title:
    for word in i.split(" "):
        words_counts[word] += 1

print('There are', len(words_counts) ,'words in the job title column')
words_counts.most_common()

There are 183 words in the job title column


[('human', 39),
 ('resources', 39),
 ('aspire', 13),
 ('professional', 10),
 ('seek', 10),
 ('management', 7),
 ('manager', 7),
 ('university', 6),
 ('student', 6),
 ('business', 5),
 ('generalist', 5),
 ('specialist', 4),
 ('position', 4),
 ('senior', 3),
 ('opportunities', 3),
 ('director', 3),
 ('major', 3),
 ('college', 2),
 ('graduate', 2),
 ('english', 2),
 ('coordinator', 2),
 ('officer', 2),
 ('internship', 2),
 ('retail', 2),
 ('staff', 2),
 ('north', 2),
 ('america', 2),
 ('information', 2),
 ('systems', 2),
 ('leader', 2),
 ('bauer', 1),
 ('magna', 1),
 ('cum', 1),
 ('laude', 1),
 ('native', 1),
 ('teacher', 1),
 ('epik', 1),
 ('program', 1),
 ('korea', 1),
 ('people', 1),
 ('development', 1),
 ('ryan', 1),
 ('advisory', 1),
 ('board', 1),
 ('member', 1),
 ('celal', 1),
 ('bayar', 1),
 ('humber', 1),
 ('system', 1),
 ('chapman', 1),
 ('vice', 1),
 ('president', 1),
 ('chief', 1),
 ('market', 1),
 ('communications', 1),
 ('corporate', 1),
 ('social', 1),
 ('responsibility', 1

We can infer that the most frequent words in the job title column are: 'human', 'resources', 'aspire', 'professional' and 'seek'.

#3. Modeling part
 3.1. Rank candidates based on a given keyword
#### Applying Word Embedding Techniques

In natural language processing (NLP), word embedding is used to represent words mathematically for text analysis, typically in the form of a real-valued vector that encodes the meaning of the word, such that the words that are closer in the vector space are expected to be similar in meaning. Indeed, the words need to be made meaningful for machine learning or deep learning algorithms. They must be expressed numerically. Algorithms such as TF-IDF, Word2Vec, Doc2vec, Glove, FastText, Transformed models like Bert enable words/sentences to be expressed mathematically as word embedding techniques used to solve such problems.

##1. Apply tfidf
TF-IDF (term frequency-inverse document frequency) is a statistical measure that evaluates how relevant a word is to a document in a collection of documents.

In [ ]:
df_1=df.copy()

In [ ]:
def rank_candidates_tfidf(keyword,df,feature_name):
    #store cleaned rows into a list
    corpus=df[feature_name].tolist()
    #apply preprocessing on keyword
    keyword=preprocess_text(keyword)
    #create an instance of tfidfVectorizer
    tfvectoriser = TfidfVectorizer()
    #Append the cleaned keyword to th corpus list
    corpus.append(keyword)
    #fit the Tfidf vectorizer on the cleaned data
    key_tf=tfvectoriser.fit_transform(corpus)
    #create a dataframe that contains tfidf values of each token for each row in th data
    key_df=pd.DataFrame(key_tf.toarray(), columns=tfvectoriser.get_feature_names_out())
    shape=key_df.shape
    #calculate the cosine similarity between the keyword and the data's rows
    cosine_sim = cosine_similarity(key_tf.toarray()[:shape[0]-1], key_tf.toarray()[shape[0]-1].reshape(1, -1))

    #Add cosine resluts into a new column fit
    df['fit_tfidf']=cosine_sim
    #store the ranked candidates based on the fit score
    final_df=df.sort_values('fit_tfidf',ascending=False)
    #remove the keyword added to reuse the function again on other different keywords
    corpus.pop()
    return final_df
    #GLOVE and Fasttex

Test the rank_candidate function

In [ ]:
keyword='aspire human resources'
df_1=rank_candidates_tfidf(keyword,df_1,"job_title")

In [ ]:
#number of ranked candidates based on the keyword
print('There are',df_1[df_1['fit_tfidf']!=0].shape[0],'ranked candidates for the job',keyword)

There are 35 ranked candidates for the job aspire human resources


From the reslut below, we can infer

In [ ]:
#Explore the new column
df_1.head(10)

job_title  \
id                                                       
97                 aspire human resources professional   
3                  aspire human resources professional   
6                    aspire human resources specialist   
73   aspire human resources manager seek internship...   
27   aspire human resources management student seek...   
72   business management major aspire human resourc...   
66   experience retail manager aspire human resourc...   
74                        human resources professional   
7    student humber college aspire human resources ...   
100  aspire human resources manager graduate may se...   

                                location connection  fit_tfidf  
id                                                              
97                  Kokomo, Indiana Area         71   0.757836  
3    Raleigh-Durham, North Carolina Area         44   0.757836  
6             Greater New York City Area          1   0.665663  
73                   Houston, Texas Area          7   0.601381  
27                   Houston, Texas Area      500+    0.428562  
72                Monroe, Louisiana Area          5   0.426162  
66                    Austin, Texas Area         57   0.396419  
74                   Greater Boston Area         16   0.393191  
7                                 Kanada         61   0.382504  
100             Cape Girardeau, Missouri        103   0.345520

In [ ]:
df_1.shape

(53, 4)

#2- Apply word2vec
Word2vec is another of the frequently used word embedding techniques. The entire corpus is scanned, and the vector creation process is performed by determining which words the target word occurs with more often. In this way, the semantic closeness of the words to each other is also revealed.
There are two kinds of word2vec: CBOW (continuous bag of words model) and Skipgram (diff is how you train word embedding). We will use then the first type. We have used zero vector to implement word2 vector.

In [ ]:
#function to get tokens and the word2vec model
def gettokens_word2vec(keyword,df,feature_name):
    #store cleaned rows into a list
    corpus=df[feature_name].tolist()
    #apply preprocessing on keyword
    keyword=preprocess_text(keyword)
    #Append the cleaned keyword to th corpus list
    corpus.append(keyword)
    #create tokens from our corpus
    corpus_tokens = [word_tokenize(line) for line in corpus]
    #create an instance of word2vec
    w = w2v(corpus_tokens, min_count=3, window=5)# min_count=2 means consider just sentenses with more than 2 words
    #build vocabs from the corpus
    w.build_vocab(corpus_tokens, progress_per=1000)
    #Get top 10 similar words to 'resources'
    print('Top 10 similar words to the word "resources":',w.wv.most_similar('resources', topn=10))
    return corpus_tokens, w

#define keyword
keyword='aspire human resources'
corpus_tokens, w = gettokens_word2vec(keyword,df_1,'job_title')

Top 10 similar words to the word "resources": [('student', 0.19914473593235016), ('generalist', 0.17272962629795074), ('specialist', 0.17018413543701172), ('seek', 0.1460377275943756), ('manager', 0.06404177844524384), ('position', 0.04653667286038399), ('management', -0.0027353083714842796), ('professional', -0.013374509289860725), ('university', -0.016800248995423317), ('director', -0.01963910646736622)]


In [ ]:
#function to join tokens with average of each sentense
def sentense_vector(corpus,w):
    line_vec=np.array(0)
    l=list()
    for i in corpus:
        a=0
        for j in i:
            try:
                a=a+1
                line_vec=line_vec+w.wv[j]
            except:
                line_vec=line_vec
        l.append(line_vec/a)
    return l

In [ ]:
corpus_tokens[:2]

[['aspire', 'human', 'resources', 'professional'],
 ['aspire', 'human', 'resources', 'professional']]

In [ ]:
#final main function
def rank_candidates_word2vec(df):
    #join tokens with average of each sentense
    all_vec=sentense_vector(corpus_tokens,w)
    #create a dataframe that contains word2vec vectors
    vec_df=pd.DataFrame(all_vec)
    shape=vec_df.shape
    # Explore keys created with theire indexs
    w.wv.key_to_index
    #calculate the cosine similarity between the keyword and the data's rows
    cosine_sim = cosine_similarity(all_vec[:shape[0]-1], all_vec[shape[0]-1].reshape(1, -1))
    #Add cosine resluts into a new column fit
    df['fit_word2vec']=cosine_sim
    #store the ranked candidates based on the fit score
    final_df=df.sort_values('fit_word2vec',ascending=False)
    #remove the keyword added to reuse the function again on other different keywords
    corpus_tokens.pop()
    return final_df

In [ ]:
df_1=rank_candidates_word2vec(df_1)

In [ ]:
#number of ranked candidates based on the keyword
print('There are',df_1[df_1['fit_word2vec']!=0].shape[0],'ranked candidates for the job',keyword)

There are 53 ranked candidates for the job aspire human resources


As we can see below, the final result is stored in the fit_word2vec. As we can see, results are not good because word2vec tries to look at words and doesn't take care of the global context. Thus, let's see other techniques that looks also for the context of the word in the sentence.

In [ ]:
df_1

job_title  \
id                                                       
104             director administration excellence log   
80             junior mes engineer information systems   
85   rrp brand portfolio executive jti japan tobacc...   
86   information systems specialist programmer love...   
87   bachelor science biology victoria university w...   
90     undergraduate research assistant styczynski lab   
2    native english teacher epik english program korea   
92   seek employment opportunities within customer ...   
93   admissions representative community medical ce...   
91           lead official western illinois university   
4                  people development coordinator ryan   
96   student indiana university kokomo business man...   
103                                 always set success   
98                                             student   
11                          student chapman university   
102          business intelligence analytics travelers   
95                  student westfield state university   
5         advisory board member celal bayar university   
75   nortia staff seek human resources payroll admi...   
77   human resources conflict management policies p...   
69   director human resources north america groupe ...   
81   senior human resources business partner heil e...   
70   retire army national guard recruiter office ma...   
83   human resources manager endemol shine north am...   
12   senior vice president chief human resources of...   
94   seek human resources opportunities open travel...   
84   human resources professional world leader gi s...   
89                         director human resources ey   
67          human resources staff recruit professional   
13   human resources coordinator intercontinental b...   
71          human resources generalist scottmadden inc   
68                human resources specialist luxottica   
78                   human resources generalist schwan   
101                 human resources generalist loparex   
8                    human resources senior specialist   
28                  seek human resources opportunities   
88                    human resources management major   
99                       seek human resources position   
76   aspire human resources professional passionate...   
1    bauer college business graduate magna cum laud...   
10   seek human resources human resources managemen...   
100  aspire human resources manager graduate may se...   
79   liberal arts major aspire human resources analyst   
82   aspire human resources professional energetic ...   
27   aspire human resources management student seek...   
7    student humber college aspire human resources ...   
74                        human resources professional   
72   business management major aspire human resourc...   
73   aspire human resources manager seek internship...   
66   experience retail manager aspire human resourc...   
6                    aspire human resources specialist   
3                  aspire human resources professional   
97                 aspire human resources professional   

                                location connection  fit_tfidf  fit_word2vec  
id                                                                            
104                          Katy, Texas      500+    0.000000      0.999920  
80     Myrtle Beach, South Carolina Area         52   0.000000      0.999754  
85             Greater Philadelphia Area      500+    0.000000      0.999754  
86                Gaithersburg, Maryland          4   0.000000      0.999754  
87                   Baltimore, Maryland         40   0.000000      0.999653  
90                  Greater Atlanta Area        155   0.000000      0.999460  
2                                 Kanada      500+    0.000000      0.999460  
92                  Torrance, California         64   0.000000      0.999460  
93                Long Beach, California          9   0.000000  

In [ ]:
df_1.shape

(53, 5)

#3. Glove
Glove stands for the acronym Global Vector. It focuses on words co-occurrences over the whole corpus. Its embeddings relate to the probabilities that two words appear together. It Consider the global property of the data.\
First, we should convert the GloVe file containing the word embeddings to the word2vec format for convenience of use. We can do it using the gensim library, a function called glove2word2vec.

In [ ]:
!wget http://nlp.stanford.edu/data/glove.6B.zip
!unzip glove.6B.zip

--2024-09-16 20:27:30--  http://nlp.stanford.edu/data/glove.6B.zip
Resolving nlp.stanford.edu (nlp.stanford.edu)... 171.64.67.140
Connecting to nlp.stanford.edu (nlp.stanford.edu)|171.64.67.140|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://nlp.stanford.edu/data/glove.6B.zip [following]
--2024-09-16 20:27:30--  https://nlp.stanford.edu/data/glove.6B.zip
Connecting to nlp.stanford.edu (nlp.stanford.edu)|171.64.67.140|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://downloads.cs.stanford.edu/nlp/data/glove.6B.zip [following]
--2024-09-16 20:27:31--  https://downloads.cs.stanford.edu/nlp/data/glove.6B.zip
Resolving downloads.cs.stanford.edu (downloads.cs.stanford.edu)... 171.64.64.22
Connecting to downloads.cs.stanford.edu (downloads.cs.stanford.edu)|171.64.64.22|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 862182613 (822M) [application/zip]
Saving to: ‘glove.6B.zip.1’

gl

In [ ]:
!unzip glove.6B.zip

Archive:  glove.6B.zip
replace glove.6B.50d.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [ ]:

# Specify file paths
glove_input_file = "glove.6B.100d.txt"
word2vec_output_file = "glove"+'.word2vec'


In [ ]:
# Convert GloVe format to Word2Vec format
glove2word2vec(glove_input_file, word2vec_output_file)

(400000, 100)

In [ ]:
#function to get tokens and the word2vec model
def gettokens_glove2vec(keyword,df,feature_name,glove_path):
    #store cleaned rows into a list
    corpus=df[feature_name].tolist()
    #apply preprocessing on keyword
    keyword=preprocess_text(keyword)
    #Append the cleaned keyword to th corpus list
    corpus.append(keyword)
    #create tokens from our corpus
    corpus_tokens = [word_tokenize(line) for line in corpus]
    #create an instance of Glove model
    glove2vec_file = glove_path
    model_glove = KeyedVectors.load_word2vec_format(glove2vec_file, binary=False)
    #get vector of a word
    model_glove.get_vector('king')[:3]
    #Get top 10 similar words to 'resources'
    print('similar words to the word "aspire":',model_glove.similar_by_word('aspire'))
    return corpus_tokens, model_glove

#define keyword
keyword='aspire human resources'
corpus_tokens, glove = gettokens_glove2vec(keyword,df_1,'job_title',"glove.word2vec")

similar words to the word "aspire": [('yearn', 0.645106852054596), ('aspires', 0.6096683144569397), ('aspired', 0.6053599715232849), ('strive', 0.600728452205658), ('pretend', 0.5369201898574829), ('forsake', 0.5339478850364685), ('someday', 0.526309609413147), ('covet', 0.5244596600532532), ('emulate', 0.5235080718994141), ('realise', 0.5204852819442749)]


In [ ]:
corpus_tokens[:3]

[['aspire', 'human', 'resources', 'professional'],
 ['aspire', 'human', 'resources', 'professional'],
 ['aspire', 'human', 'resources', 'specialist']]

In [ ]:
#function to join tokens with average of each sentense
def sentense_vector_glove(corpus,glove):
    line_vec=np.array(0)
    l=list()
    for i in corpus:
        a=0
        for j in i:
            try:
                a=a+1
                line_vec=line_vec+glove.get_vector(j)
            except:
                line_vec=line_vec
        l.append(line_vec/a)
    return l

In [ ]:
#final main function
def rank_candidates_glove(df):
    #join tokens with average of each sentense
    all_vec_glove=sentense_vector_glove(corpus_tokens,glove)
    #create a dataframe that contains word2vec vectors
    vec_df_glove=pd.DataFrame(all_vec_glove)
    shape=vec_df_glove.shape
    #calculate the cosine similarity between the keyword and the data's rows
    cosine_sim = cosine_similarity(all_vec_glove[:shape[0]-1], all_vec_glove[shape[0]-1].reshape(1, -1))
    #Add cosine resluts into a new column fit
    df['fit_glove']=cosine_sim
    #store the ranked candidates based on the fit score
    final_df=df.sort_values('fit_glove',ascending=False)
    #remove the keyword added to reuse the function again on other different keywords
    corpus_tokens.pop()
    return final_df

In [ ]:
df_1=rank_candidates_glove(df_1)

In [ ]:
#number of ranked candidates based on the keyword
print('There are',df_1[df_1['fit_glove']!=0].shape[0],'ranked candidates for the job',keyword)

There are 53 ranked candidates for the job aspire human resources


In [ ]:
df_1.head(7)

job_title  \
id                                                       
103                                 always set success   
85   rrp brand portfolio executive jti japan tobacc...   
93   admissions representative community medical ce...   
2    native english teacher epik english program korea   
95                  student westfield state university   
87   bachelor science biology victoria university w...   
96   student indiana university kokomo business man...   

                       location connection  fit_tfidf  fit_word2vec  fit_bert  \
id                                                                              
103    Greater Los Angeles Area      500+         0.0      0.997614  0.032193   
85    Greater Philadelphia Area      500+         0.0      0.999754  0.088577   
93       Long Beach, California          9        0.0      0.999264  0.145111   
2                        Kanada      500+         0.0      0.999460  0.154695   
95   Bridgewater, Massachusetts         57        0.0      0.995624  0.174335   
87          Baltimore, Maryland         40        0.0      0.999653  0.190097   
96           Lafayette, Indiana         19        0.0      0.998742  0.190871   

     fit_glove  
id              
103   0.999981  
85    0.999953  
93    0.999855  
2     0.999789  
95    0.999600  
87    0.999413  
96    0.998978

#4 - Apply Fastext

In [ ]:
!wget https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.en.300.bin.gz

--2024-09-17 03:18:37--  https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.en.300.bin.gz
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 13.35.7.38, 13.35.7.82, 13.35.7.128, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|13.35.7.38|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4503593528 (4.2G) [application/octet-stream]
Saving to: ‘cc.en.300.bin.gz’

cc.en.300.bin.gz    100%[===================>]   4.19G  24.2MB/s    in 1m 51s  

2024-09-17 03:20:29 (38.6 MB/s) - ‘cc.en.300.bin.gz’ saved [4503593528/4503593528]



In [ ]:
!gunzip cc.en.300.bin.gz
#use python fastex library. pypi distribution.whitout downloading the model.
#move Bert part after glove and fastex.

In [ ]:
fast_model = fasttext.util.reduce_model(fasttext.load_model('cc.en.300.bin'), 100)

In [ ]:
#fast=KeyedVectors.load_word2vec_format("cc.en.300.vec", binary=False)

In [ ]:
# df=data.copy()
# df['job_title'] = df['job_title'].apply(preprocess_text)
# corpus=df['job_title'].tolist()
# keyword='aspire human resources'
# #corpus.append(keyword)
# corpus_tokens = [word_tokenize(line) for line in corpus]

In [ ]:
# result = fast.most_similar(positive=['woman', 'king'], negative=['man'], topn=1)
# print('Most similar word to King + Woman: ', result)

In [ ]:
# join tokens with average of each sentense
# def sentense_vector_fast(corpus):
#     line_vec=np.array(0)
#     l=list()
#     for i in corpus:
#         a=0
#         for j in i:
#             try:
#                 a=a+1
#                 line_vec=line_vec+fast.get_vector(j)
#             except:
#                 line_vec=line_vec
#         l.append(line_vec/a)
#     return l

In [ ]:
# all_vec_fast=sentense_vector_fast(corpus_tokens)
# vec_df_fast=pd.DataFrame(all_vec_fast)
# vec_df_fast.shape

In [ ]:
# cosine_sim = cosine_similarity(all_vec_fast[:53], all_vec_fast[53].reshape(1, -1))
# df['fit_fast']=cosine_sim
# final_df=df[df['fit_fast']!=0].sort_values('fit_fast',ascending=False)
# final_df.head(10)

#5. Apply Bert
BERT stands for Bidirectional Encoder Representations from Transformers. It is designed to pre-train deep bidirectional representations from unlabeled text by jointly conditioning on both left and right context. As a result, the pre-trained BERT model can be fine-tuned with just one additional output layer to create state-of-the-art models for a wide range of NLP tasks.

In [ ]:
#Load the pretrained model
bert = SentenceTransformer('all-MiniLM-L6-v2')

In [ ]:
def rank_candidates_bert(keyword,df,feature_name):
    #load bert model
    bert = SentenceTransformer('all-MiniLM-L6-v2')
    #store cleaned rows into a list
    corpus=df[feature_name].tolist()
    #apply preprocessing on keyword
    keyword=preprocess_text(keyword)
    #Append the cleaned keyword to th corpus list
    corpus.append(keyword)
    #encode our sentences
    sentence_embeddings = bert.encode(corpus)
    print("the shape of our encoded sentences is",sentence_embeddings.shape)
    shape=sentence_embeddings.shape
    #calculate the cosine similarity between the keyword and the data's rows
    cosine_sim = cosine_similarity(sentence_embeddings[:shape[0]-1], sentence_embeddings[shape[0]-1].reshape(1, -1))
    #Add cosine resluts into a new column fit
    df['fit_bert']=cosine_sim
    #store the ranked candidates based on the fit score
    final_df=df.sort_values('fit_bert',ascending=False)
    #remove the keyword added to reuse the function again on other different keywords
    corpus.pop()
    return final_df

In [ ]:
keyword='aspire human resources'
df_1=rank_candidates_bert(keyword,df_1,"job_title")

the shape of our encoded sentences is (54, 384)


In [ ]:
#number of ranked candidates based on the keyword
print('There are',df_1[df_1['fit_bert']!=0].shape[0],'ranked candidates for the job',keyword)

There are 53 ranked candidates for the job aspire human resources


In [ ]:
#Explore the new column
df_1.head(10)

job_title  \
id                                                       
97                 aspire human resources professional   
3                  aspire human resources professional   
6                    aspire human resources specialist   
74                        human resources professional   
66   experience retail manager aspire human resourc...   
73   aspire human resources manager seek internship...   
100  aspire human resources manager graduate may se...   
72   business management major aspire human resourc...   
79   liberal arts major aspire human resources analyst   
7    student humber college aspire human resources ...   

                                location connection  fit_tfidf  fit_word2vec  \
id                                                                             
97                  Kokomo, Indiana Area         71   0.757836      0.852264   
3    Raleigh-Durham, North Carolina Area         44   0.757836      0.852264   
6             Greater New York City Area          1   0.665663      0.896906   
74                   Greater Boston Area         16   0.393191      0.958898   
66                    Austin, Texas Area         57   0.396419      0.949685   
73                   Houston, Texas Area          7   0.601381      0.953347   
100             Cape Girardeau, Missouri        103   0.345520      0.967693   
72                Monroe, Louisiana Area          5   0.426162      0.954558   
79           Baton Rouge, Louisiana Area          7   0.342890      0.964966   
7                                 Kanada         61   0.382504      0.959331   

     fit_bert  
id             
97   0.928787  
3    0.928787  
6    0.916763  
74   0.718712  
66   0.708162  
73   0.707531  
100  0.696832  
72   0.695717  
79   0.695704  
7    0.678677

From the result above, we can see clearly that Bert is the best model till now. And this is not arbitrary, as Bert is pretrained on a large corpus of unlabelled text including the entire Wikipedia(that’s 2,500 million words!) and Book Corpus (800 million words). Indeed, pre-training step is half the magic behind BERT’s success.

End of the part 1 of our project:
 I was able to apply word embeeding thechniques to make the first ranked list. The following step is to rerank this list when a candidate is starred at the top.

##3.2. Re-rank the list whene some candidates are starred
As the Bert model was the best, so we will use it for this step. In this step, we will use the learning to rank (LTR) technique, a machine learning sub-field applicable to a variety of real-world problems that are related to ranking prediction or candidate recommendation. There are 3 main known models: RankNet, LambdaRank and LmabdaMart. In our wase, we will try to use the RankNet model first and then LamdaRank.

In [ ]:
df_1["starred"]=df_1["fit_bert"]

In [ ]:
starred_id = [int(item) for item in input("Please, inter ids of candidates you want to star?: ").split()]

Please, inter ids of candidates you want to star?: 6 42


In [ ]:
for id in starred_id:
    df_1['starred'][id]=1

In [ ]:
df_1.head(10)

job_title  \
id                                                       
103                                 always set success   
85   rrp brand portfolio executive jti japan tobacc...   
93   admissions representative community medical ce...   
2    native english teacher epik english program korea   
95                  student westfield state university   
87   bachelor science biology victoria university w...   
96   student indiana university kokomo business man...   
98                                             student   
102          business intelligence analytics travelers   
5         advisory board member celal bayar university   

                       location connection  fit_tfidf  fit_word2vec  fit_bert  \
id                                                                              
103    Greater Los Angeles Area      500+         0.0      0.997614  0.032193   
85    Greater Philadelphia Area      500+         0.0      0.999754  0.088577   
93       Long Beach, California          9        0.0      0.999264  0.145111   
2                        Kanada      500+         0.0      0.999460  0.154695   
95   Bridgewater, Massachusetts         57        0.0      0.995624  0.174335   
87          Baltimore, Maryland         40        0.0      0.999653  0.190097   
96           Lafayette, Indiana         19        0.0      0.998742  0.190871   
98          Houston, Texas Area          4        0.0      0.997614  0.191936   
102  Greater New York City Area         49        0.0      0.995984  0.203006   
5                İzmir, Türkiye      500+         0.0      0.993605  0.208498   

     fit_glove   starred  
id                        
103   0.999981  0.032193  
85    0.999953  0.088577  
93    0.999855  0.145111  
2     0.999789  0.154695  
95    0.999600  0.174335  
87    0.999413  0.190097  
96    0.998978  0.190871  
98    0.998258  0.191936  
102   0.998101  0.203006  
5     0.998001  0.208498

## 3.2.1. RankNet

In [ ]:
# Implement the Ranknet model
class RankNet(nn.Module):
    def __init__(self, inputs=384, outputs=1):
        super(RankNet, self).__init__()
        self.model = nn.Sequential(
            nn.Linear(inputs, 256),
            nn.Dropout(0.5),
            #nn.ReLU(inplace=True),
            nn.Linear(256, 128),
            nn.Dropout(0.5),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(128, outputs),
            nn.Sigmoid()
        )
        self.sigmoid = nn.Sigmoid()

    def forward(self, input_1, input_2):
        result_1 = self.model(input_1)
        result_2 = self.model(input_2)
        pred = self.sigmoid(result_1 - result_2) #input_1 is more relevant than input_2
        return pred

    def predict(self, input):
        result = self.model(input)
        return result

In [ ]:
#GENERATE DATA
rows_1 = df_1.sample(n = 100, replace = True)
rows_2 = df_1.sample(n = 100, replace = True)

In [ ]:
#get list of job titles for each data generated
job_list_1 = list(rows_1['job_title'])
job_list_2 = list(rows_2['job_title'])

In [ ]:
doc_1 = bert.encode(job_list_1)
doc_2 = bert.encode(job_list_2)
doc_1 = torch.from_numpy(doc_1).float()
doc_2 = torch.from_numpy(doc_2).float()

In [ ]:
print('doc_1.shape',doc_1.shape)
print('doc_2.shape',doc_2.shape)

doc_1.shape torch.Size([100, 384])
doc_2.shape torch.Size([100, 384])


In [ ]:
#Define Y true labels
y_1_true = list(rows_1['starred'])
y_2_true = list(rows_2['starred'])
y_true = torch.tensor([1.0 if y1_i>y2_i else 0.5 if y1_i==y2_i else 0.0 for y1_i, y2_i in zip(y_1_true, y_2_true)]).float()

In [ ]:
y_true=y_true.unsqueeze(1)

In [ ]:
y_true.shape

torch.Size([100, 1])

In [ ]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def train(optimiz, list_lrs, num_epochs):

    dict_best=dict()
    dict_hidden=dict()
    loss_best=0
    num_epochs = num_epochs
    model = RankNet().to(device)
    #Loss function and optimizer
    criterion = nn.BCELoss()
    for lr in list_lrs:
        if optimiz == 'SGD':
            optimizer = torch.optim.SGD(model.parameters(), lr = lr, momentum = 0.9)
        elif optimiz == 'Adam':
            optimizer = torch.optim.Adam(model.parameters(), lr = lr)
        elif optimiz =='Adadelta':
            optimizer = torch.optim.Adadelta(model.parameters(), lr = lr)
        elif optimiz =='ADAMW':
            optimizer = torch.optim.AdamW(model.parameters(), lr = lr)

        print('lr: ', lr, 'optimizer: ', optimiz)
        base_path = os.path.abspath(os.path.join(os.getcwd(), '..'))
        base_path = os.path.dirname(base_path)
        data_path = base_path + '/train_result.txt'

        total_step = len(y_true)
        for epoch in range(num_epochs):
            pred = model(doc_1, doc_2)
            loss = criterion(pred, y_true)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            if epoch % 100 == 0:
              print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch + 1, num_epochs, loss.item()))
            if loss_best==0:
                loss_best=loss.item()
            if loss.item()<loss_best:
                loss_best=loss.item()
                name_model=optimiz+'model.ckpt'
                torch.save(model.state_dict(), name_model)

        dict_hidden[lr]=loss_best
        print('>' * 60)
    dict_best[optimiz]=dict_hidden
    lr_min=min(dict_hidden, key=dict_hidden.get)
    print("=="*100)
    print('Best model with the optimizer', optimiz,':learning rate =:',lr_min,'; loss =', dict_hidden[lr_min])
    return model, loss

In [ ]:
#Test 4 types of optimizers and different learning rates
optimizer_list=['Adam','SGD','Adadelta','ADAMW']
for optimiz in optimizer_list:
    model, loss =train(optimiz=optimiz, list_lrs=[0.2, 0.1, 0.01, 0.001, 0.0001, 0.00001], num_epochs=1000)

lr:  0.2 optimizer:  Adam
Epoch [1/1000], Loss: 0.6931
Epoch [101/1000], Loss: 0.5728
Epoch [201/1000], Loss: 0.7144
Epoch [301/1000], Loss: 0.6786
Epoch [401/1000], Loss: 0.5526
Epoch [501/1000], Loss: 0.6648
Epoch [601/1000], Loss: 0.6806
Epoch [701/1000], Loss: 0.6734
Epoch [801/1000], Loss: 0.6598
Epoch [901/1000], Loss: 0.6848
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
lr:  0.1 optimizer:  Adam
Epoch [1/1000], Loss: 0.6972
Epoch [101/1000], Loss: 0.6784
Epoch [201/1000], Loss: 0.5582
Epoch [301/1000], Loss: 0.6752
Epoch [401/1000], Loss: 0.6888
Epoch [501/1000], Loss: 0.6931
Epoch [601/1000], Loss: 0.6931
Epoch [701/1000], Loss: 0.6931
Epoch [801/1000], Loss: 0.6931
Epoch [901/1000], Loss: 0.6931
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
lr:  0.01 optimizer:  Adam
Epoch [1/1000], Loss: 0.6931
Epoch [101/1000], Loss: 0.6931
Epoch [201/1000], Loss: 0.6931
Epoch [301/1000], Loss: 0.6931
Epoch [401/1000], Loss: 0.6931
Epoch [501/1000], Loss: 0.6931

Load the best model which is the model whith ADAMW otimizer and a learning rate equal to 0.2



In [ ]:
#load th ebest model to our device
model = RankNet().to(device)
model.load_state_dict(torch.load('ADAMWmodel.ckpt'))
#ADAMW optimization
#Lamda Rank
#lang chain framework for utilizing LLM (after bert)
#create a promt to rank employees.

<All keys matched successfully>

In [ ]:
pred_scores = []
for i in df_1['job_title']:
    sentence_embeddings = bert.encode(i)
    sentence_embeddings_tensor = torch.from_numpy(sentence_embeddings).float()
    pred = round(model.predict(sentence_embeddings_tensor).detach().numpy().sum(),2)
    pred_scores.append(pred)

df_1['rerank_fit'] = pred_scores
df_1.sort_values(by ='rerank_fit', ascending = False).head()

job_title  \
id                                                      
69  director human resources north america groupe ...   
82  aspire human resources professional energetic ...   
68               human resources specialist luxottica   
81  senior human resources business partner heil e...   
88                   human resources management major   

                               location connection  fit_tfidf  fit_word2vec  \
id                                                                            
69  Greater Grand Rapids, Michigan Area      500+    0.139734      0.991564   
82                   Austin, Texas Area        174   0.324746      0.962783   
68           Greater New York City Area      500+    0.220481      0.987559   
81          Chattanooga, Tennessee Area        455   0.141075      0.991187   
88                 Milpitas, California         18   0.250501      0.978463   

    fit_bert  fit_glove   starred  rerank_fit  
id                                             
69  0.583638   0.983542  0.583638         1.0  
82  0.662319   0.978967  0.662319         1.0  
68  0.570694   0.983193  0.570694         1.0  
81  0.528106   0.983153  0.528106         1.0  
88  0.607106   0.982635  0.607106         1.0

Let's see the total dataframe.

In [ ]:
df_1.sort_values(by ='rerank_fit', ascending = False)

job_title  \
id                                                       
69   director human resources north america groupe ...   
82   aspire human resources professional energetic ...   
68                human resources specialist luxottica   
81   senior human resources business partner heil e...   
88                    human resources management major   
8                    human resources senior specialist   
101                 human resources generalist loparex   
13   human resources coordinator intercontinental b...   
27   aspire human resources management student seek...   
76   aspire human resources professional passionate...   
83   human resources manager endemol shine north am...   
7    student humber college aspire human resources ...   
99                       seek human resources position   
10   seek human resources human resources managemen...   
79   liberal arts major aspire human resources analyst   
28                  seek human resources opportunities   
78                   human resources generalist schwan   
72   business management major aspire human resourc...   
100  aspire human resources manager graduate may se...   
73   aspire human resources manager seek internship...   
66   experience retail manager aspire human resourc...   
74                        human resources professional   
6                    aspire human resources specialist   
3                  aspire human resources professional   
67          human resources staff recruit professional   
97                 aspire human resources professional   
84   human resources professional world leader gi s...   
70   retire army national guard recruiter office ma...   
1    bauer college business graduate magna cum laud...   
75   nortia staff seek human resources payroll admi...   
89                         director human resources ey   
94   seek human resources opportunities open travel...   
77   human resources conflict management policies p...   
85   rrp brand portfolio executive jti japan tobacc...   
96   student indiana university kokomo business man...   
102          business intelligence analytics travelers   
98                                             student   
95                  student westfield state university   
87   bachelor science biology victoria university w...   
92   seek employment opportunities within customer ...   
2    native english teacher epik english program korea   
93   admissions representative community medical ce...   
5         advisory board member celal bayar university   
80             junior mes engineer information systems   
91           lead official western illinois university   
86   information systems specialist programmer love...   
11                          student chapman university   
104             director administration excellence log   
90     undergraduate research assistant styczynski lab   
4                  people development coordinator ryan   
71          human resources generalist scottmadden inc   
12   senior vice president chief human resources of...   
103                                 always set success   

                                location connection  fit_tfidf  fit_word2vec  \
id                                                                             
69   Greater Grand Rapids, Michigan Area      500+    0.139734      0.991564   
82                    Austin, Texas Area        174   0.324746      0.962783   
68            Greater New York City Area      500+    0.220481      0.987559   
81           Chattanooga, Tennessee Area        455   0.141075      0.991187   
88                  Milpitas, California         18   0.250501      0.978463   
8                 San Francisco Bay Area      500+    0.240780      0.983178   
101  Raleigh-Durham, North Carolina Area      500+    0.224491      0.984853   
13                      Atlanta, Georgia      500+    0.148562      0.989463   
27                   Houston, Texas Area      500+    

##3.2.2. LambdaRank

In [ ]:
y_true.shape

torch.Size([100, 1])

In [ ]:
# Define the LambdaRank model
class LambdaRank(nn.Module):
    def __init__(self, input_size=384, output_size=1):
        super(LambdaRank, self).__init__()
        self.model = nn.Sequential(
            nn.Linear(input_size, 256),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(128, output_size)
        )

    def forward(self, input_1, input_2):
        # Calculate the score for input_1
        score_1 = self.model(input_1)
        # Calculate the score for input_2
        score_2 = self.model(input_2)
        return score_1, score_2  # Return two separate scores (pred_1, pred_2)

    def predict(self, input):
        return self.model(input)  # Predicts a single input's score


NameError: name 'nn' is not defined

In [ ]:
# Define LambdaRank loss function (fixed)
def lambda_rank_loss(score_1, score_2, y_true_1, y_true_2):
    """
    Custom loss function for LambdaRank.
    score_1: Predicted score for input_1 (tensor)
    score_2: Predicted score for input_2 (tensor)
    y_true_1: True relevance score for input_1 (tensor)
    y_true_2: True relevance score for input_2 (tensor)
    """
    # Ensure y_true_1 and y_true_2 are tensors
    y_true_1 = torch.tensor(y_true_1).to(score_1.device)  # Ensure same device as the model
    y_true_2 = torch.tensor(y_true_2).to(score_2.device)

    # Compute pairwise difference in scores
    S = (y_true_1 > y_true_2).float() - (y_true_1 < y_true_2).float()  # Relevance pair (1, -1, 0)

    # RankNet-style loss (pairwise log loss)
    pred_diff = score_1 - score_2

    # Ensure S has the same shape as pred_diff
    S = S.unsqueeze(1) # add an extra dimension to S

    lambda_rank_loss = F.binary_cross_entropy_with_logits(pred_diff, S)

    return lambda_rank_loss

In [ ]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def train(optimiz, list_lrs, num_epochs):

    dict_best = dict()
    dict_hidden = dict()
    loss_best = 0
    num_epochs = num_epochs
    model = LambdaRank().to(device)  # Initialize the LambdaRank model
    # Remove the BCELoss() as we're replacing it with lambda_rank_loss

    for lr in list_lrs:
        if optimiz == 'SGD':
            optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
        elif optimiz == 'Adam':
            optimizer = torch.optim.Adam(model.parameters(), lr=lr)
        elif optimiz == 'Adadelta':
            optimizer = torch.optim.Adadelta(model.parameters(), lr=lr)
        elif optimiz == 'ADAMW':
            optimizer = torch.optim.AdamW(model.parameters(), lr=lr)

        print('lr: ', lr, 'optimizer: ', optimiz)
        base_path = os.path.abspath(os.path.join(os.getcwd(), '..'))
        base_path = os.path.dirname(base_path)
        data_path = base_path + '/train_result.txt'

        total_step = len(y_true)
        for epoch in range(num_epochs):
            # Forward pass: model prediction for both sets of inputs (doc_1 and doc_2)
            pred_1, pred_2 = model(doc_1, doc_2)

            # Compute LambdaRank loss using the two predictions and the true labels
            loss = lambda_rank_loss(pred_1, pred_2, y_1_true, y_2_true)

            # Backward pass and optimization
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            if epoch % 100 == 0:
                print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch + 1, num_epochs, loss.item()))

            # Keep track of the best loss and save the model
            if loss_best == 0:
                loss_best = loss.item()
            if loss.item() < loss_best:
                loss_best = loss.item()
                name_model = optimiz + 'model.ckpt'
                torch.save(model.state_dict(), name_model)

        dict_hidden[lr] = loss_best
        print('>' * 60)
    dict_best[optimiz] = dict_hidden
    lr_min = min(dict_hidden, key=dict_hidden.get)
    print("=="*100)
    print('Best model with the optimizer', optimiz, ': learning rate =:', lr_min, '; loss =', dict_hidden[lr_min])

    return model, loss


In [ ]:
#Test 4 types of optimizers and different learning rates
optimizer_list=['Adam','SGD','Adadelta','ADAMW']
for optimiz in optimizer_list:
    model, loss =train(optimiz=optimiz, list_lrs=[0.2, 0.1, 0.01, 0.001, 0.0001, 0.00001], num_epochs=1000)

lr:  0.2 optimizer:  Adam
Epoch [1/1000], Loss: 0.6935
Epoch [101/1000], Loss: 0.6931
Epoch [201/1000], Loss: 0.6893
Epoch [301/1000], Loss: 0.6993
Epoch [401/1000], Loss: 0.6931
Epoch [501/1000], Loss: 0.6931
Epoch [601/1000], Loss: 0.6931
Epoch [701/1000], Loss: 0.6931
Epoch [801/1000], Loss: 0.6931
Epoch [901/1000], Loss: 0.6931
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
lr:  0.1 optimizer:  Adam
Epoch [1/1000], Loss: 0.6931
Epoch [101/1000], Loss: 0.6931
Epoch [201/1000], Loss: 0.6931
Epoch [301/1000], Loss: 0.6931
Epoch [401/1000], Loss: 0.6931
Epoch [501/1000], Loss: 0.6931
Epoch [601/1000], Loss: 0.6931
Epoch [701/1000], Loss: 0.6931
Epoch [801/1000], Loss: 0.6931
Epoch [901/1000], Loss: 0.6931
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
lr:  0.01 optimizer:  Adam
Epoch [1/1000], Loss: 0.6931
Epoch [101/1000], Loss: 0.6931
Epoch [201/1000], Loss: 0.6931
Epoch [301/1000], Loss: 0.6931
Epoch [401/1000], Loss: 0.6931
Epoch [501/1000], Loss: 0.6931

In [ ]:
#load the best model to our device
model = LambdaRank().to(device)

# Load the state_dict, but set strict=False to ignore missing/unexpected keys
model.load_state_dict(torch.load('ADAMWmodel.ckpt'), strict=False)

# Print model to check if it loaded correctly
print(model)

LambdaRank(
  (model): Sequential(
    (0): Linear(in_features=384, out_features=256, bias=True)
    (1): ReLU()
    (2): Dropout(p=0.5, inplace=False)
    (3): Linear(in_features=256, out_features=128, bias=True)
    (4): ReLU()
    (5): Dropout(p=0.5, inplace=False)
    (6): Linear(in_features=128, out_features=1, bias=True)
  )
)


In [ ]:
pred_scores = []
for i in df_1['job_title']:
    sentence_embeddings = bert.encode(i)
    sentence_embeddings_tensor = torch.from_numpy(sentence_embeddings).float()
    pred = round(model.predict(sentence_embeddings_tensor).detach().numpy().sum(),2)
    pred_scores.append(pred)

df_1['rerank_fit'] = pred_scores
df_1.sort_values(by ='rerank_fit', ascending = False).head()

job_title  \
id                                                      
83  human resources manager endemol shine north am...   
12  senior vice president chief human resources of...   
77  human resources conflict management policies p...   
67         human resources staff recruit professional   
10  seek human resources human resources managemen...   

                     location connection  fit_tfidf  fit_word2vec  fit_bert  \
id                                                                            
83    Los Angeles, California        268   0.143631      0.991047  0.636403   
12        Houston, Texas Area      500+    0.124627      0.990815  0.492106   
77     Dallas/Fort Worth Area        409   0.104003      0.991913  0.442761   
67  Jackson, Mississippi Area      500+    0.192671      0.989467  0.606238   
10  Greater Philadelphia Area      500+    0.302694      0.974012  0.575650   

    fit_glove   starred  rerank_fit  
id                                   
83   0.979331  0.636403        2.32  
12   0.988212  0.492106        1.80  
77   0.993046  0.442761        1.63  
67   0.983378  0.606238        1.25  
10   0.983524  0.575650        0.93

In [ ]:
df_1.sort_values(by ='rerank_fit', ascending = False)

job_title  \
id                                                       
83   human resources manager endemol shine north am...   
12   senior vice president chief human resources of...   
77   human resources conflict management policies p...   
67          human resources staff recruit professional   
10   seek human resources human resources managemen...   
82   aspire human resources professional energetic ...   
87   bachelor science biology victoria university w...   
81   senior human resources business partner heil e...   
94   seek human resources opportunities open travel...   
90     undergraduate research assistant styczynski lab   
74                        human resources professional   
75   nortia staff seek human resources payroll admi...   
5         advisory board member celal bayar university   
95                  student westfield state university   
93   admissions representative community medical ce...   
99                       seek human resources position   
85   rrp brand portfolio executive jti japan tobacc...   
69   director human resources north america groupe ...   
13   human resources coordinator intercontinental b...   
27   aspire human resources management student seek...   
4                  people development coordinator ryan   
101                 human resources generalist loparex   
2    native english teacher epik english program korea   
11                          student chapman university   
80             junior mes engineer information systems   
70   retire army national guard recruiter office ma...   
104             director administration excellence log   
71          human resources generalist scottmadden inc   
6                    aspire human resources specialist   
86   information systems specialist programmer love...   
73   aspire human resources manager seek internship...   
1    bauer college business graduate magna cum laud...   
97                 aspire human resources professional   
103                                 always set success   
100  aspire human resources manager graduate may se...   
89                         director human resources ey   
88                    human resources management major   
8                    human resources senior specialist   
72   business management major aspire human resourc...   
98                                             student   
78                   human resources generalist schwan   
96   student indiana university kokomo business man...   
92   seek employment opportunities within customer ...   
76   aspire human resources professional passionate...   
84   human resources professional world leader gi s...   
79   liberal arts major aspire human resources analyst   
91           lead official western illinois university   
66   experience retail manager aspire human resourc...   
68                human resources specialist luxottica   
28                  seek human resources opportunities   
102          business intelligence analytics travelers   
7    student humber college aspire human resources ...   
3                  aspire human resources professional   

                                location connection  fit_tfidf  fit_word2vec  \
id                                                                             
83               Los Angeles, California        268   0.143631      0.991047   
12                   Houston, Texas Area      500+    0.124627      0.990815   
77                Dallas/Fort Worth Area        409   0.104003      0.991913   
67             Jackson, Mississippi Area      500+    0.192671      0.989467   
10             Greater Philadelphia Area      500+    0.302694      0.974012   
82                    Austin, Texas Area        174   0.324746      0.962783   
87                   Baltimore, Maryland         40   0.000000      0.999653   
81           Chattanooga, Tennessee Area        455   0.141075      0.991187   
94           Amerika Birleşik Devletleri        415   

#Final Result:
As we can see above, the ranked list based on Bert model is being reranked successfully. As we can see, the 3 chosen candidates with id equal to 73 and 6 are rightly labeled as 1. Then, we can say that our model reaches strongly the goal of this project.

#USE Large Language Model
This code installs the necessary libraries and sets up OpenAI's GPT-3.5 for ranking employee job titles based on relevance to a specific keyword. A PromptTemplate is used to create a structured prompt asking the GPT-3.5 model to rank employees by their job titles and relevance to the provided keyword. The llm object sends the formatted prompt to the GPT-3.5 model, which generates the ranking result. You can replace "Your_OpenAI_Key" with your actual OpenAI API key to run this code with GPT-3.5.

In [ ]:
!pip install langchain --upgrade # make sure to upgrade langchain to the latest version
!pip install langchain-community
!pip install openai

from langchain.llms import OpenAI
from langchain.prompts import PromptTemplate

# Replace with your OpenAI API key
os.environ["OPENAI_API_KEY"] = "Your_OpenAI_Key"

llm = OpenAI(model="gpt-3.5-turbo", temperature=0)


template = """
You are a expert in ranking employees based on their job titles and relevance to a specific keyword.
Given the following list of employees and their job titles, and a keyword, rank them in order of relevance to the keyword.

Employees:
{employees}

Keyword:
{keyword}

Ranking Criteria:
- How closely the job title matches the keyword.
- The overall relevance of the employee's job title to the search query.

Output:
A numbered list of employees ranked by relevance, starting with the most relevant.
"""

prompt = PromptTemplate(
    input_variables=["employees", "keyword"],
    template=template,
)

# Example usage
employees = df_1['job_title'].tolist()
keyword = 'aspire human resources'
formatted_prompt = prompt.format(employees=employees, keyword=keyword)
ranking_result = llm(formatted_prompt)

print(ranking_result)
#Add explanation
# use 'job_title'

RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

#Ranking Employees Using GPT-2 with Hugging Face Pipeline
This code uses Hugging Face’s Mixtral-8x7B-Instruct-v0.1 interface api model to rank employee job titles based on relevance to a specific keyword. It defines a prompt template that takes a list of employee job titles and a keyword, formats it, and generates a ranking using the Hugging Face pipeline for text generation. The model is explicitly set to truncate the input and pad appropriately. The dataframe data is used to extract the top 20 job titles for the ranking task.

In [4]:
# Install necessary packages
!pip install langchain-community
!pip install transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 30.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 30.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.9/400.9 kB 17.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 292.2/292.2 kB 14.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.3/49.3 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.9/141.9 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 3.8 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 9.0.0
    Uninstalling tenacity-9.0.0:
      Successfully uninstalled tenacity-9.0.0


In [64]:
import time
import requests
from langchain.prompts import PromptTemplate

# Replace 'your_token' with your Hugging Face API token
API_URL = "https://api-inference.huggingface.co/models/mistralai/Mixtral-8x7B-Instruct-v0.1"
headers = {"Authorization": f"Bearer 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"}

# Create the prompt template
template = """
You are an expert in ranking candidates based on their job titles and relevance to a specific keyword.
Given the following list of candidates job titles and the search term rank them using cosine similarity. The output should be a numbered list of candidates ranked by relevance, starting with the most relevant.
You should just list the candidates in {candidates} that is most similar to {keyword}. Do not search on other jobs on internet.

candidates:
{candidates}

search term:
{keyword}
"""

prompt = PromptTemplate(
    input_variables=["candidates", "keyword"],
    template=template,
)

# Example data: List of candidates and keyword
candidates = candidates = df['job_title'].head(20).tolist()
keyword = 'Aspiring human resources'

# Format the prompt with candidate data and keyword
formatted_prompt = prompt.format(candidates=candidates, keyword=keyword)

# Retry loop to handle model loading
retry = True
while retry:
    response = requests.post(API_URL, headers=headers, json={"inputs": formatted_prompt})
    response_json = response.json()

    if response.status_code == 503:
        print(f"Model is loading, estimated time: {response_json['estimated_time']} seconds")
        time.sleep(response_json['estimated_time'])  # Wait for the estimated time before retrying
    elif "error" in response_json:
        print(f"Error: {response_json['error']}")
        retry = False  # Stop retrying if there's an error other than loading
    else:
        # Successfully received the generated text
        generated_text = response_json[0]["generated_text"]
        print(generated_text)
        retry = False



You are an expert in ranking candidates based on their job titles and relevance to a specific keyword.
Given the following list of candidates job titles and the search term rank them using cosine similarity. The output should be a numbered list of candidates ranked by relevance, starting with the most relevant.
You should just list the candidates in ['bauer college business graduate magna cum laude aspire human resources professional', 'native english teacher epik english program korea', 'aspire human resources professional', 'people development coordinator ryan', 'advisory board member celal bayar university', 'aspire human resources specialist', 'student humber college aspire human resources generalist', 'human resources senior specialist', 'seek human resources human resources management system generalist position', 'student chapman university', 'senior vice president chief human resources officer market communications corporate social responsibility officer engie houston woodlands

In [36]:
import time
import requests
from langchain.prompts import PromptTemplate

# Replace 'your_token' with your Hugging Face API token
API_URL = "https://api-inference.huggingface.co/models/mistralai/Mixtral-8x7B-Instruct-v0.1"
headers = {"Authorization": f"Bearer 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"}

# Create the prompt template
template = """
which one of {candidates} are similar to {keyword}
"""

prompt = PromptTemplate(
    input_variables=["candidates", "keyword"],
    template=template,
)

# Example data: List of candidates and keyword
candidates = candidates = ["Data Scientist", "Software Engineer", "HR Specialist", "Marketing Manager", "Human resources"]
keyword = 'Aspiring human resources'

# Format the prompt with candidate data and keyword
formatted_prompt = prompt.format(candidates=candidates, keyword=keyword)

# Retry loop to handle model loading
retry = True
while retry:
    response = requests.post(API_URL, headers=headers, json={"inputs": formatted_prompt})
    response_json = response.json()

    if response.status_code == 503:
        print(f"Model is loading, estimated time: {response_json['estimated_time']} seconds")
        time.sleep(response_json['estimated_time'])  # Wait for the estimated time before retrying
    elif "error" in response_json:
        print(f"Error: {response_json['error']}")
        retry = False  # Stop retrying if there's an error other than loading
    else:
        # Successfully received the generated text
        generated_text = response_json[0]["generated_text"]
        print(generated_text)
        retry = False



which one of ['Data Scientist', 'Software Engineer', 'HR Specialist', 'Marketing Manager', 'Human resources'] are similar to "Aspiring human resources"

Answer:

['HR Specialist', 'Human resources'] are similar to "Aspiring human resources". 'HR Specialist' is a professional who specializes in human resources management, and 'Human resources' refers to the department within an organization that is responsible for managing the workforce. Both of these options are relevant to the field of human resources, which is what "Aspiring human resources" is referring to.



#Conclusion:
From the steps that I followed in this project, I can infer that BERT model was the best to find the similarity between our data and the targeted keyword/phrase. As for the second part of Re-Ranking, I run the based model of LTR techniques which is the RankNet model. The best model was with the optimizer SGD and a learning rate equal to 0.2. Thus, gives a loss score of 0.44%.